### PR TITLE
Upgrade to http-body 1.0

### PR DESCRIPTION
**Pull request by davidpdrsn**
Still some things to do but should allow https://github.com/tokio-rs/axum/pull/1882 to move forward.

## TODO

- [x] Publish new version of http-body that has `Frame::map_data`
- [x] Are we okay with having http-body-util in our public API? We have things like https://docs.rs/tower-http/latest/tower_http/limit/struct.RequestBodyLimit.html#impl-Service%3CRequest%3CReqBody%3E%3E-for-RequestBodyLimit%3CS%3E which includes `Limited`. Should we create wrappers?
- [x] Update (de)compression, but how?
- [x] Update docs
- [x] Update examples. This is blocked on axum, tonic, and warp supporting hyper 1.0. Examples are commented out for now until axum, tonic, etc have been updated.
- [x] Changelog

### CODE REVIEW

**Comment by rakshith-ravi** [comment]
Hey. New to tower-http codebase. Complex tasks would certainly take me time to understand and work on. Are there any low-hanging fruits that I can help with, and then I can slowly work my way up?

Just want to see if there's any way I can help push this ahead

**Comment by davidpdrsn** [comment]
Yay everything (almost) builds!

The only thing left to port is compression/decompression. I have commented out that code for now while I got everything else working. I'll start looking at compression/decompression soon.

**Comment by jplatte** [comment]
@davidpdrsn you said on the axum issue for hyper 1.0 that you want to release on the same day hyper 1.0 comes out, in that case can you have a look at #368 before then? I think it might be the only PR that is a breaking change to the API that's currently open, and it's very small.

**Comment by jplatte** [comment]
http-body 1.0 is out!

**Comment by davidpdrsn** [comment]
Yes! I'm gonna update this and the axum PR tomorrow.

File: examples/axum-key-value-store/src/main.rs [hunk 1]
```diff
@@ -1,3 +1,8 @@
+fn main() {
+    eprint!("this example has not yet been updated to hyper 1.0");
```
Comment (jplatte) [inline examples/axum-key-value-store/src/main.rs:2]:
```suggestion
    eprintln!("this example has not yet been updated to hyper 1.0");
```

😜

File: tower-http/src/classify/status_in_range_is_error.rs [hunk 2]
```diff
@@ -4,34 +4,6 @@ use std::{fmt, ops::RangeInclusive};
 
 /// Response classifier that considers responses with a status code within some range to be
 /// failures.
-///
-/// # Example
-///
-/// A client with tracing where server errors _and_ client errors are considered failures.
```
Comment (jplatte) [inline tower-http/src/classify/status_in_range_is_error.rs:10]: Why did you remove this example?

**Review by jplatte — COMMENTED** [review]
18 / 59 files viewed

File: tower-http/src/classify/status_in_range_is_error.rs (hunk 2 above)
Comment (davidpdrsn) [reply tower-http/src/classify/status_in_range_is_error.rs:10]: I could find a way to make a hyper client that implemented `tower::Service` but I see now that hyper-util provides one! I've brought this example and the one in `lib.rs` back.

**Comment by davidpdrsn** [comment]
I've managed to port compression and decompression 🎉 So there should be no more blockers!

**Comment by davidpdrsn** [comment]
@GlenDC has volunteered to help reviewing this 🙏

File: tower-http/src/compression_utils.rs [hunk 3]
```diff
@@ -275,8 +296,63 @@ where
...
-        self.project().body.poll_data(cx)
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            let this = self.as_mut().project();
+
+            if *this.yielded_all_data {
+                return Poll::Ready(None);
+            }
+
+            match futures_util::ready!(this.body.poll_frame(cx)) {
```
Comment (jplatte) [inline tower-http/src/compression_utils.rs:307]: `use std::task::ready;`?

File: tower-http/src/compression_utils.rs [hunk 4]
```diff
@@ -275,8 +296,63 @@ where
...
+    B: Body,
+{
+    type Data = B::Data;
+    type Error = B::Error;
+
+    fn poll_frame(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
+        // First drive the stream impl. This conume all data frames and buffer at most one trailers
```
Comment (jplatte) [inline tower-http/src/compression_utils.rs:335]:
```suggestion
        // First drive the stream impl. This consumes all data frames and buffer at most one trailers
```

File: tower-http/src/compression_utils.rs [hunk 5]
```diff
@@ -352,6 +428,7 @@ pub enum CompressionLevel {
     Precise(i32),
 }
 
+#[allow(clippy::derivable_impls)]
```
Comment (jplatte) [inline tower-http/src/compression_utils.rs:431]: Why not derive this? Clippy should take the MSRV into account, indeed deriving default on an enum (where the default variant has no fields) requires only [Rust 1.62](https://blog.rust-lang.org/2022/06/30/Rust-1.62.0.html#default-enum-variants).

File: tower-http/src/lib.rs [hunk 6]
```diff
@@ -17,17 +17,18 @@
 //! ```rust,no_run
 //! use tower_http::{
 //!     add_extension::AddExtensionLayer,
-//!     compression::CompressionLayer,
+//!     // compression::CompressionLayer,
```
Comment (jplatte) [inline tower-http/src/lib.rs:20]: Should this still be commented out?

File: tower-http/src/lib.rs [hunk 7]
```diff
@@ -64,7 +65,7 @@
 //!         // Share an `Arc<State>` with all requests
 //!         .layer(AddExtensionLayer::new(Arc::new(state)))
 //!         // Compress responses
-//!         .layer(CompressionLayer::new())
+//!         // .layer(CompressionLayer::new())
```
Comment (jplatte) [inline tower-http/src/lib.rs:68]: related

**Review by jplatte — APPROVED** [review]
Some more small comments, have now reviewed completely and looks good to me overall (obviously 😄).

File: examples/axum-key-value-store/src/main.rs [hunk 8]
```diff
@@ -1,3 +1,8 @@
+fn main() {
+    eprintln!("this example has not yet been updated to hyper 1.0");
```
Comment (GlenDC) [inline examples/axum-key-value-store/src/main.rs:2]: Why not? I did it with little effort for the `tower-async` version. Want me to contribute this as a PR to your PR?

File: examples/tonic-key-value-store/src/main.rs [hunk 9]
```diff
@@ -1,3 +1,8 @@
+fn main() {
+    eprint!("this example has not yet been updated to hyper 1.0");
```
Comment (GlenDC) [inline examples/tonic-key-value-store/src/main.rs:2]: Why not? I did it with little effort for the `tower-async` version. Want me to contribute this as a PR to your PR?

File: tower-http/src/body.rs [hunk 10]
```diff
@@ -0,0 +1,121 @@
+//! Body types.
+//!
+//! All these are wrappers around other body types. You shouldn't have to use them in your code.
+//! Use `http-body-util` instead.
+//!
+//! They exist because we don't want to expose types from `http-body-util` in `tower-http`s public
```
Comment (GlenDC) [inline tower-http/src/body.rs:6]: Why not?

File: tower-http/src/catch_panic.rs [hunk 11]
```diff
@@ -341,7 +349,7 @@ impl ResponseForPanic for DefaultResponseForPanic {
             );
         };
 
-        let mut res = Response::new(Full::from("Service panicked"));
+        let mut res = Response::new(Full::new(http_body_util::Full::from("Service panicked")));
```
Comment (GlenDC) [inline tower-http/src/catch_panic.rs:352]: You anyway make use of it 🤷 Might as well just re-rexpose it then if you really want to ? Not sure I see the value. I imagine it has to do with stability guarantees, but dunno... Would hold stronger the argument if you do not link into it at all.

File: tower-http/src/body.rs (hunk 10 above)
Comment (jplatte) [reply tower-http/src/body.rs:6]: The futures crates, including `futures-util`, are very stable. `http-body-util` is much more likely to see breaking-change releases soon-ish, and if we had it as a public dependency, it would require a breaking-change release of `tower-http` to upgrade to the latest version.

File: tower-http/src/catch_panic.rs (hunk 11 above)
Comment (davidpdrsn) [reply tower-http/src/catch_panic.rs:352]: What jplatte where here is the reason https://github.com/tower-rs/tower-http/pull/348#discussion_r1399789699

Using stuff from http-body-util internally is fine but we won't expose those types in our public API.

File: examples/tonic-key-value-store/src/main.rs (hunk 9 above)
Comment (davidpdrsn) [reply examples/tonic-key-value-store/src/main.rs:2]: How did you port it? Tonic doesn't support http 1.0. Did you do a bunch of mapping of the types?

File: examples/axum-key-value-store/src/main.rs (hunk 8 above)
Comment (davidpdrsn) [reply examples/axum-key-value-store/src/main.rs:2]: Same reason as https://github.com/tower-rs/tower-http/pull/348#discussion_r1399815537, axum doesn't yet support http 1.0

File: tower-http/src/content_encoding.rs [hunk 12]
```diff
@@ -148,7 +148,7 @@ impl QValue {
         let mut c = s.chars();
         // Parse "q=" (case-insensitively).
         match c.next() {
-            Some('q') | Some('Q') => (),
+            Some('q' | 'Q') => (),
```
Comment (GlenDC) [inline tower-http/src/content_encoding.rs:151]: hehe you ran clippy on this codebase?

File: tower-http/src/cors/allow_private_network.rs [hunk 13]
```diff
@@ -39,6 +39,10 @@ impl AllowPrivateNetwork {
         Self(AllowPrivateNetworkInner::Predicate(Arc::new(f)))
     }
 
+    #[allow(
```
Comment (GlenDC) [inline tower-http/src/cors/allow_private_network.rs:42]: Hmmm, I wonder why I don't need these on tower-async. Are we running different versions of clippy or different configs?

File: tower-http/src/services/fs/serve_dir/tests.rs [hunk 14]
```diff
@@ -474,7 +473,7 @@ async fn read_partial_in_bounds() {
         )));
     assert_eq!(res.headers()["content-type"], "text/markdown");
 
-    let body = hyper::body::to_bytes(res.into_body()).await.ok().unwrap();
+    let body = to_bytes(res.into_body()).await.ok().unwrap();
```
Comment (GlenDC) [inline tower-http/src/services/fs/serve_dir/tests.rs:476]: FYI I import these as `crate::test_helpers::self` as I find that for functions like `to_bytes` it's a lot clearer when defined as `test_helpers::to_bytes` to know immediately where it comes from

**Review by GlenDC — APPROVED** [review]
Besides the few, mostly open or nitpick comment. All looks excellent to me.

File: examples/axum-key-value-store/src/main.rs (hunk 8 above)
Comment (GlenDC) [reply examples/axum-key-value-store/src/main.rs:2]: True. For now I was using your branch. I get you might not want to do that, but now you do not have an example at all... so dunno.

File: examples/tonic-key-value-store/src/main.rs (hunk 9 above)
Comment (GlenDC) [reply examples/tonic-key-value-store/src/main.rs:2]: Sorry I was wrong. Tonic and Warp examples I deleted during my original port as I do not use these libraries. I only kept in the hyper and axum ones and those are the only once I fixed this time as well.

File: tower-http/src/body.rs (hunk 10 above)
Comment (GlenDC) [reply tower-http/src/body.rs:6]: Understood. Something I am less concerned about. But I understand that this is for tower not an option.

File: tower-http/src/catch_panic.rs (hunk 11 above)
Comment (GlenDC) [reply tower-http/src/catch_panic.rs:352]: Agreed.

File: tower-http/src/cors/allow_private_network.rs (hunk 13 above)
Comment (davidpdrsn) [reply tower-http/src/cors/allow_private_network.rs:42]: Not sure. I really wanna cut down on the number of feature flags we use anyway. A job for another day 😅

File: tower-http/src/decompression/mod.rs [hunk 15]
```diff
@@ -189,14 +183,4 @@ mod tests {
             .insert("content-encoding", "gzip".parse().unwrap());
         Ok(res)
     }
-
-    #[allow(dead_code)]
-    async fn is_compatible_with_hyper() {
```
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply tower-http/src/decompression/mod.rs:194]: Good question! It could be brought back with `hyper_util`'s legacy Client.

**Comment by jplatte** [comment]
Hm, we do not have an example on how to combine this w/ `tower-util` to serve a `ServeDir` or `Redirect` directly without any extra routing from axum, warp or whatever?

File: examples/axum-key-value-store/src/main.rs (hunk 8 above)
Comment (Oliboy50) [reply examples/axum-key-value-store/src/main.rs:2]: > axum doesn't yet support http 1.0

but [now it does](https://tokio.rs/blog/2023-11-27-announcing-axum-0-7-0), right?

Comment (davidpdrsn) [reply examples/axum-key-value-store/src/main.rs:2]: Yep it does.

Comment (Oliboy50) [reply examples/axum-key-value-store/src/main.rs:2]: A PR to restore this example using axum 0.7: https://github.com/tower-rs/tower-http/pull/448
